SourceBufferPrivate::ResetParserState shouldn't synchronously dispatch tasks

jyavenard opened the pull request:
#### e448a8292b724c70d880956c2bc56c647070d287
<pre>
SourceBufferPrivate::ResetParserState shouldn&apos;t synchronously dispatch tasks
<a href="https://bugs.webkit.org/show_bug.cgi?id=255120">https://bugs.webkit.org/show_bug.cgi?id=255120</a>
rdar://107729258

Reviewed by Youenn Fablet.

The MSE&apos;s Reset Parser State algorithm [1] alongside the `abort()` [2]
operation is racy at best and leads to a non-deterministicc behaviour as
the buffer append algorithm started by the `appendBuffer()` [3] operation
runs asynchronously.

As such, having parsed the entire input buffer when the `abort()` method
is called is a valid behaviour, and we make this behaviour the default.

We now consider all operations to be sequential and we queue them to run
sequentially.
We actually abort and reset the parser only once the previous operations
have completed, all while allowing `appendBuffer()` to be called immediately
after `abort()`
As `abort()` use case is designed to be able to append a new buffer regardless
of the operation that was run before (such as a partial init or media segment)
and we continue to fullfill this requirement.

This allows to remove the need to actually abort a currently running append
operation.
This will also help when we implement MSE in a worker as the possibility
of the race allowed by the spec does no longer exists.

[1] <a href="https://w3c.github.io/media-source/#sourcebuffer-reset-parser-state">https://w3c.github.io/media-source/#sourcebuffer-reset-parser-state</a>
[2] <a href="https://w3c.github.io/media-source/#dom-sourcebuffer-abort">https://w3c.github.io/media-source/#dom-sourcebuffer-abort</a>
[3] <a href="https://w3c.github.io/media-source/#dom-sourcebuffer-appendbuffer">https://w3c.github.io/media-source/#dom-sourcebuffer-appendbuffer</a>

* LayoutTests/media/media-source/media-mp4-h264-partial-abort-expected.txt:
* LayoutTests/media/media-source/media-mp4-h264-partial-abort.html: Ensure that if abort is called
then `update` event isn&apos;t fired.

* LayoutTests/media/media-source/media-mp4-h264-partial-abort-expected.txt:
* LayoutTests/media/media-source/media-mp4-h264-partial-abort.html:
* Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:
(WebCore::SourceBufferPrivate::advanceOperationState):
(WebCore::SourceBufferPrivate::rewindOperationState):
(WebCore::SourceBufferPrivate::appendCompleted):
(WebCore::SourceBufferPrivate::processOperation):
(WebCore::SourceBufferPrivate::didReceiveInitializationSegment):
(WebCore::SourceBufferPrivate::didReceiveSample):
(WebCore::SourceBufferPrivate::append):
(WebCore::SourceBufferPrivate::queueOperation):
(WebCore::SourceBufferPrivate::processPendingOperations):
(WebCore::SourceBufferPrivate::abortPendingOperations):
(WebCore::SourceBufferPrivate::processError):
(WebCore::SourceBufferPrivate::abort):
(WebCore::SourceBufferPrivate::resetParserState):
(WebCore::SourceBufferPrivate::processInitOperation): Deleted.
(WebCore::SourceBufferPrivate::processMediaSamples): Deleted.
* Source/WebCore/platform/graphics/SourceBufferPrivate.h:
* Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.mm:
(WebCore::SourceBufferPrivateAVFObjC::~SourceBufferPrivateAVFObjC):
(WebCore::SourceBufferPrivateAVFObjC::didParseInitializationData):
(WebCore::SourceBufferPrivateAVFObjC::appendInternal):
(WebCore::SourceBufferPrivateAVFObjC::abort):
(WebCore::SourceBufferPrivateAVFObjC::resetParserStateInternal):
(WebCore::SourceBufferPrivateAVFObjC::append): Deleted.
(WebCore::SourceBufferPrivateAVFObjC::resetParserState): Deleted.
* Source/WebCore/platform/graphics/gstreamer/mse/SourceBufferPrivateGStreamer.cpp:
(WebCore::SourceBufferPrivateGStreamer::appendInternal):
(WebCore::SourceBufferPrivateGStreamer::resetParserStateInternal):
(WebCore::SourceBufferPrivateGStreamer::append): Deleted.
(WebCore::SourceBufferPrivateGStreamer::abort): Deleted.
(WebCore::SourceBufferPrivateGStreamer::resetParserState): Deleted.
* Source/WebCore/platform/graphics/gstreamer/mse/SourceBufferPrivateGStreamer.h:
* Source/WebCore/platform/mock/mediasource/MockSourceBufferPrivate.cpp:
(WebCore::MockSourceBufferPrivate::appendInternal):
(WebCore::MockSourceBufferPrivate::resetParserStateInternal):
(WebCore::MockSourceBufferPrivate::append): Deleted.
(WebCore::MockSourceBufferPrivate::abort): Deleted.
* Source/WebCore/platform/mock/mediasource/MockSourceBufferPrivate.h:
* Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemote.cpp:
(WebKit::SourceBufferPrivateRemote::appendInternal):
(WebKit::SourceBufferPrivateRemote::resetParserStateInternal):
* Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemote.h:

Canonical link: <a href="https://commits.webkit.org/262746@main">https://commits.webkit.org/262746@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a396740f5d8d71dfaac0343dbd6f583064023c17

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/2424 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/2424 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/2544 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/3693 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/2467 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/2394 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/2568 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/2512 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/2152 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/2446 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/2205 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/2178 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/3517 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/180 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/2163 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/2032 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/2289 "Built successfully") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/2191 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/3351 "Passed tests") | 
| | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/2213 "Passed tests") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/2009 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/2214 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/2165 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/611 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/2182 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/2338 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:
```diff
@@ -617,65 +639,75 @@ void SourceBufferPrivate::didReceiveSample(Ref<MediaSample>&& sample)
 void SourceBufferPrivate::abortPendingOperations()
 {
-    for (auto& operation : m_pendingAppendOperations) {
+    for (auto& operation : m_pendingOperations) {
```
std::exchange(m_pendingOperations, { }) to remove the clear method?

jyavenard replied:
done

youennf commented on Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.mm:
```diff
@@ -706,18 +697,16 @@ static void bufferWasConsumedCallback(CMNotificationCenterRef, const void* liste
-    m_appendQueue->dispatchSync([] { });
-    m_parser->resetParserState();
-    SourceBufferPrivate::resetParserState();
+    m_appendQueue->dispatch([parser = m_parser] {
```
Maybe keep ensuring that m_parser is not destroyed in m_appendQueue by dispatching a task with a ref to m_parser that gets back to main thread.

jyavenard replied:
I made SourceBufferParser inherits from ThreadSafeRefCounted<SourceBufferParser, WTF::DestructionThread::Main>

youennf commented on Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:
```diff
@@ -617,65 +639,75 @@ void SourceBufferPrivate::didReceiveSample(Ref<MediaSample>&& sample)
+        auto operation = m_pendingOperations.takeFirst();
         std::visit(WTF::makeVisitor([&](InitOperation&& initOperation) {
-            processInitOperation(WTFMove(initOperation));
+            processOperation(WTFMove(initOperation));
```
I tend to prefer more specific names like processInitOperation, it helps for searching

jyavenard replied:
re-renamed the methods

jyavenard committed:
SourceBufferPrivate::ResetParserState shouldn't synchronously dispatch tasks
https://bugs.webkit.org/show_bug.cgi?id=255120
rdar://107729258

Reviewed by Youenn Fablet.

The MSE's Reset Parser State algorithm [1] alongside the `abort()` [2]
operation is racy at best and leads to a non-deterministicc behaviour as
the buffer append algorithm started by the `appendBuffer()` [3] operation
runs asynchronously.

As such, having parsed the entire input buffer when the `abort()` method
is called is a valid behaviour, and we make this behaviour the default.

We now consider all operations to be sequential and we queue them to run
sequentially.
We actually abort and reset the parser only once the previous operations
have completed, all while allowing `appendBuffer()` to be called immediately
after `abort()`
As `abort()` use case is designed to be able to append a new buffer regardless
of the operation that was run before (such as a partial init or media segment)
and we continue to fullfill this requirement.

This allows to remove the need to actually abort a currently running append
operation.
This will also help when we implement MSE in a worker as the possibility
of the race allowed by the spec does no longer exists.

[1] https://w3c.github.io/media-source/#sourcebuffer-reset-parser-state
[2] https://w3c.github.io/media-source/#dom-sourcebuffer-abort
[3] https://w3c.github.io/media-source/#dom-sourcebuffer-appendbuffer

* LayoutTests/media/media-source/media-mp4-h264-partial-abort-expected.txt:
* LayoutTests/media/media-source/media-mp4-h264-partial-abort.html: Ensure that if abort is called
then `update` event isn't fired.

* LayoutTests/media/media-source/media-mp4-h264-partial-abort-expected.txt:
* LayoutTests/media/media-source/media-mp4-h264-partial-abort.html:
* Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:
(WebCore::SourceBufferPrivate::advanceOperationState):
(WebCore::SourceBufferPrivate::rewindOperationState):
(WebCore::SourceBufferPrivate::appendCompleted):
(WebCore::SourceBufferPrivate::processOperation):
(WebCore::SourceBufferPrivate::didReceiveInitializationSegment):
(WebCore::SourceBufferPrivate::didReceiveSample):
(WebCore::SourceBufferPrivate::append):
(WebCore::SourceBufferPrivate::queueOperation):
(WebCore::SourceBufferPrivate::processPendingOperations):
(WebCore::SourceBufferPrivate::abortPendingOperations):
(WebCore::SourceBufferPrivate::processError):
(WebCore::SourceBufferPrivate::abort):
(WebCore::SourceBufferPrivate::resetParserState):
(WebCore::SourceBufferPrivate::processInitOperation): Deleted.
(WebCore::SourceBufferPrivate::processMediaSamples): Deleted.
* Source/WebCore/platform/graphics/SourceBufferPrivate.h:
* Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.mm:
(WebCore::SourceBufferPrivateAVFObjC::~SourceBufferPrivateAVFObjC):
(WebCore::SourceBufferPrivateAVFObjC::didParseInitializationData):
(WebCore::SourceBufferPrivateAVFObjC::appendInternal):
(WebCore::SourceBufferPrivateAVFObjC::abort):
(WebCore::SourceBufferPrivateAVFObjC::resetParserStateInternal):
(WebCore::SourceBufferPrivateAVFObjC::append): Deleted.
(WebCore::SourceBufferPrivateAVFObjC::resetParserState): Deleted.
* Source/WebCore/platform/graphics/gstreamer/mse/SourceBufferPrivateGStreamer.cpp:
(WebCore::SourceBufferPrivateGStreamer::appendInternal):
(WebCore::SourceBufferPrivateGStreamer::resetParserStateInternal):
(WebCore::SourceBufferPrivateGStreamer::append): Deleted.
(WebCore::SourceBufferPrivateGStreamer::abort): Deleted.
(WebCore::SourceBufferPrivateGStreamer::resetParserState): Deleted.
* Source/WebCore/platform/graphics/gstreamer/mse/SourceBufferPrivateGStreamer.h:
* Source/WebCore/platform/mock/mediasource/MockSourceBufferPrivate.cpp:
(WebCore::MockSourceBufferPrivate::appendInternal):
(WebCore::MockSourceBufferPrivate::resetParserStateInternal):
(WebCore::MockSourceBufferPrivate::append): Deleted.
(WebCore::MockSourceBufferPrivate::abort): Deleted.
* Source/WebCore/platform/mock/mediasource/MockSourceBufferPrivate.h:
* Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemote.cpp:
(WebKit::SourceBufferPrivateRemote::appendInternal):
(WebKit::SourceBufferPrivateRemote::resetParserStateInternal):
* Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemote.h:

Canonical link: https://commits.webkit.org/262746@main
